### PR TITLE
Build the verification libraries using Kani compiler

### LINKED ISSUE #2533: Build Kani library with the kani-compiler
**Proposed change:** Build Kani and the std library using kani-compiler.

**Motivation:** The Kani compiler has extra checks, such as well-formed attributes, that will be useful when compiling our sysroot. For example, if we add an ill-formed `#[kani::unstable]` attribute today, the user crate compilation will fail (we even add a `debug_assert()` since we expect that to be a Kani bug. Instead, we want the `kani` library compilation to fail.

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

This will ensure we run Kani extra validation checks in our libraries.

### Resolved issues:

Resolves #2533 

### Call-outs:

By default, the Kani compiler will not codegen anything, so this change shouldn't generate Goto. Because of that, this change reduced the size of the verification library *rlib files by almost 60MB

### Testing:

* How is this change tested?

* Is this a refactor change?

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: tools/build-kani/src/sysroot.rs [hunk 1]
```diff
@@ -56,35 +56,42 @@ pub fn kani_playback_lib() -> PathBuf {
...
 /// Build the `lib/` folder and `lib-playback/` for the new sysroot.
 /// - The `lib/` folder contains the sysroot for verification.
 /// - The `lib-playback/` folder contains the sysroot used for playback.
-pub fn build_lib() -> Result<()> {
-    build_verification_lib()?;
-    build_playback_lib()
+pub fn build_lib(bin_folder: &Path) -> Result<()> {
+    let compiler = bin_folder.join("kani-compiler");
+    build_verification_lib(&compiler)?;
+    build_playback_lib(&compiler)
```
Comment (adpaco-aws) [inline tools/build-kani/src/sysroot.rs:69]: Do we discuss this in our developer documentation? If not, I think it'd be nice to add an overview there in the future.

File: tools/build-kani/src/sysroot.rs [hunk 2]
```diff
@@ -56,35 +56,42 @@ pub fn kani_playback_lib() -> PathBuf {
...
 }
 
 /// Build the `lib/` folder for the new sysroot used during verification.
 /// This will include Kani's libraries as well as the standard libraries compiled with --emit-mir.
-fn build_verification_lib() -> Result<()> {
+fn build_verification_lib(compiler: &Path) -> Result<()> {
```
Comment (adpaco-aws) [inline tools/build-kani/src/sysroot.rs:74]: Please rename `compiler` to `compiler_path` or any other name that hints at this being a location, not an object.

File: tools/build-kani/src/sysroot.rs [hunk 3]
```diff
@@ -56,35 +56,42 @@ pub fn kani_playback_lib() -> PathBuf {
...
         ["-Z", "build-std=panic_abort,std,test", "--config", "profile.dev.panic=\"abort\""];
-    build_kani_lib(&kani_sysroot_lib(), &extra_args)
+    let compiler_args = ["--kani-compiler", "-Cllvm-args=--ignore-global-asm"];
+    build_kani_lib(compiler, &kani_sysroot_lib(), &extra_args, &compiler_args)
 }
 
 /// Build the `lib-playback/` folder that will be used during counter example playback.
 /// This will include Kani's libraries compiled with `concrete-playback` feature enabled.
-fn build_playback_lib() -> Result<()> {
+fn build_playback_lib(compiler: &Path) -> Result<()> {
```
Comment (adpaco-aws) [inline tools/build-kani/src/sysroot.rs:83]: Same renaming request here.

File: tools/build-kani/src/sysroot.rs [hunk 4]
```diff
@@ -225,18 +232,20 @@ fn build_artifacts(cargo_cmd: &mut Child) -> Vec<Artifact> {
         .collect()
 }
 
+/// Build Kani binaries with the extra arguments provided and return path to the binaries folder.
```
Comment (adpaco-aws) [inline tools/build-kani/src/sysroot.rs:235]:
```suggestion
/// Build Kani binaries with the extra arguments provided and return the path to the binaries folder.
```

**Review by adpaco-aws — CHANGES_REQUESTED** [review]

File: tools/build-kani/src/sysroot.rs (hunk 1 above)
Comment (celinval) [reply tools/build-kani/src/sysroot.rs:69]: Do you mean discuss the build process?

File: tools/build-kani/src/sysroot.rs (hunk 2 above)
Comment (celinval) [reply tools/build-kani/src/sysroot.rs:74]: Sure

File: tools/build-kani/src/sysroot.rs (hunk 1 above)
Comment (adpaco-aws) [reply tools/build-kani/src/sysroot.rs]: Not the build process itself, but the libraries which are built and their purpose.

**Review by adpaco-aws — APPROVED** [review]

File: tools/build-kani/src/sysroot.rs (hunk 1 above)
Comment (celinval) [reply tools/build-kani/src/sysroot.rs]: Sounds good. I want to eventually document Kani's architecture too.

**Pull request merged**